### ISSUE TITLE
Calculate Critical Time (as opposed to measuring it)

### ISSUE THREAD

**Issue by mikeminutillo** [Particular/ServiceControl#1779]
**tl;dr - We should caluclate critical time using queue length and throughput for endpoint monitoring**

When a message is sent, it is timestamped. When it is processed, it is timestamped again and these two timestamps are compared to calculate a measurement called Critical Time. The metrics plugin captures this and sends it to ServiceControl Monitoring for display in ServicePulse.

Critical Time is a measurement of how long it takes an individual message to get processed, starting with when it was sent. As an endpoint gets busy, Critical Times tends to increase (as the queue backs up). Critical Time going up is a good metric to use to decide when it's time to scale the endpoint out so it makes sense to track it on the monitoring dashboard.

There's a major of problem with measuring critical time and applying it to an endpoint. Critical Time is measured at the end of message processing. If you record a critical time of 10 minutes, then it means the endpoint took 10 minutes to process that one message so the endpoint is likely to respond to a new message in about 10 minutes.

Unfortunately, the data is stale. 10 minutes ago, the endpoint had a critical time of 10 minutes. You have no idea if the situation right now is better or worse. There may have been a lot more messages added to the queue after this one, or there may be none. That is, the metric tells you about the historical state of the queue, it doesn't tell you anything about the state of the queue now. And the higher the critical time is (the more critical it is) the more stale it is. A critical time of 1 hour is a dire situation but the data is woefully out of date.

In the context of a single message, this is the best that you can do. In the context of a whole queue, you can do something better. This is because of two other measurements we are taking: Queue Length and Throughput. If we know there are 200 messages in the input queue and we are currently processing 10 messages per second, then we can easily figure out that the current critical time of the endpoint is 200 / 10 = 20 seconds. That is, at a rate of 10 messages per second, we will empty this queue of 200 messages in 20 seconds. This is an instantaneous measurement of the endpoint's critical time. It assumes that the processing time and throughput are fairly stable. 

As an example, imagine that we threw 1000 messages at a single endpoint at once, and it is able to process 4 per second. The first 4 messages get processed pretty quickly but after that, there are messages that needed to wait in the input queue to get processed and critical time starts to increase. When the last message is processed, it has a critical time of 250 seconds (just over 4 minutes). Someone monitoring the endpoint in ServicePulse will see the Critical Time steadily going up to 250 seconds and may decide to scale the endpoint out to try and deal with the load.

This is not an accurate representation of what is happening. The endpoint is not steadily getting less responsive over time. It is instantly getting less responsive and then getting more responsive as time passes. As soon as that initial spike of messages is added to the input queue, the critical time for the endpoint jumps to 250 seconds. Once the endpoint starts processing those messages, the critical time for the endpoint trends steadily down towards 0. An operator looking at this graph may decide not to scale out the endpoint as the system appears to be catching up and 4 minutes is an acceptable amout of latency for that endpoint.

**Comment by mikeminutillo** [comment]
To avoid confusion, it may make sense to not use the term Critical Time for this. Perhaps something like Endpoint Response Time would be better.

**Comment by ramonsmits** [comment]
That is why critical time isn't really interesting but only shows its value in combination with an SLA. 

Knowing the system currently can't cope with the load, if the load isn't decreasing we know that in X minutes/hours the system can't meet its SLA.

This is currently always an extrapolation on indeed stale data.

I think this could only be improved upon if the endpoint instance would report their amount of 'sends' as we previously did to calculate the queue length.

Yes, we could have:

- currently throughput rate
- queue length
- queue expand/deflate rate
- estimate when the backlog is gone based on throughput rate, queue expand/deflate rate and queue length.

More interesting is that the individual metrics should be plotted too.

Also, I wish we had a way we had a flexible metrics capture system. This way both our customers and ourselves could extend monitoring with all types of metrics. Business, application, and system metrics.

This way, metrics can feed other metrics

So I would say, build the foundation first to flexibly add metrics both in endpoint instances as in SC and then any of such metrics could be added in minutes.
